modules/gateway/peers.go

```diff
@@ -109,27 +120,53 @@ func (g *Gateway) randomInboundPeer() (modules.NetAddress, error) {
 	return "", errNoPeers
 }
 
-// listen handles incoming connection requests. If the connection is accepted,
+// threadedListen handles incoming connection requests. If the connection is accepted,
 // the peer will be added to the Gateway's peer list.
-func (g *Gateway) listen() {
+func (g *Gateway) threadedListen() {
+	if g.threads.Add() != nil {
+		return
+	}
+	defer g.threads.Done()
+
+	// Shutdown the listener when the stop signal is received.
+	if g.threads.Add() != nil {
+		return
+	}
+	go func() {
+		defer g.threads.Done()
+		<-g.threads.StopChan()
+		if err := g.listener.Close(); err != nil {
+			g.log.Printf("WARN: listener.Close failed: %v", err)
+		}
+	}()
```

Comment:
The `g.listener.Close()` could be moved to `gateway.Close`, before the `g.threads.Stop()`. The question is: should `g.threads.Stop()` be the sole way of stopping things? If so, we need to do things like the above. If not, we can perform other shutdown-related activity in `Close` methods. My preference is the latter; it's less surprising.